File: ts/src/map/data-type.ts
import {zodToRules, zodToTypeString} from "../internal/zod-schema";
import type {HerculesTranslation} from "../types";
import type {DataTypeClass, HerculesDataType} from "../models/data-type";

export const dataTypeMap = (klass: DataTypeClass): HerculesDataType => {
    const identifier: string = Reflect.getMetadata('hercules:identifier', klass);
    const name: HerculesTranslation[] = Reflect.getMetadata('hercules:name', klass) || [];
    const displayMessage: HerculesTranslation[] = Reflect.getMetadata('hercules:display_message', klass) || [];
    const alias: HerculesTranslation[] = Reflect.getMetadata('hercules:alias', klass) || [];
    const linkedDataTypes: string[] = Reflect.getMetadata('hercules:linked_data_type_identifiers', klass) || [];
    const genericKeys: string[] = Reflect.getMetadata('hercules:generic_keys', klass) || [];

    if (!identifier) throw new Error(`Data type class ${klass.name} is missing an identifier. Add @Identifier("your_identifier") to the class.`);

    const schema = Reflect.getMetadata('hercules:schema', klass);
    if (!schema) throw new Error(`Data type class ${klass.name} is missing a schema. Add @Schema(z.string()) to the class.`);

    return {
        identifier,
        type: zodToTypeString(schema),
        rules: zodToRules(schema),
        name,
        displayMessage,
        alias,
        linkedDataTypes,
        genericKeys,
    };
};

File: py/hercules/map/datatype.py
"""Data type mapping (port of ``src/map/datatype.map.ts``)."""
from __future__ import annotations

from hercules._metadata import get_metadata
from hercules.internal.schema import register_schema
from hercules.models.datatype import DataTypeProps


def data_type_map(klass: type) -> DataTypeProps:
    identifier = get_metadata("hercules:identifier", klass)
    name = get_metadata("hercules:name", klass) or []
    display_message = get_metadata("hercules:display_message", klass) or []
    alias = get_metadata("hercules:alias", klass) or []
    generic_keys = get_metadata("hercules:generic_keys", klass) or []

    if not identifier:
        raise ValueError(
            f"Data type class {klass.__name__} is missing an identifier. "
            f'Add @Identifier("your_identifier") to the class.'
        )

    schema = get_metadata("hercules:schema", klass)
    if schema is None:
        raise ValueError(
            f"Data type class {klass.__name__} is missing a schema. "
            f"Add @Schema(MyModel) with a Pydantic model to the class."
        )

    register_schema(schema, identifier)

    type_string = get_metadata("hercules:type_string", klass)

    return DataTypeProps(
        identifier=identifier,
        schema=schema,
        name=name,
        display_message=display_message,
        alias=alias,
        generic_keys=generic_keys,
        type_string=type_string,
    )
